File: ios/Sources/RynBridgePushFCM/PushFCMModule.swift
import Foundation
import RynBridge

public struct PushFCMModule: BridgeModule, Sendable {
    public let name = "push-fcm"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: FCMPushProvider) {
        actions = [
            "getToken": { _ in
                let token = try await provider.getToken()
                return FCMToken(token: token).toPayload()
            },
            "deleteToken": { _ in
                try await provider.deleteToken()
                return [:]
            },
            "subscribeToTopic": { payload in
                guard case let .string(topic) = payload["topic"] else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: topic")
                }
                try await provider.subscribeToTopic(topic)
                return [:]
            },
            "unsubscribeFromTopic": { payload in
                guard case let .string(topic) = payload["topic"] else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: topic")
                }
                try await provider.unsubscribeFromTopic(topic)
                return [:]
            },
            "getAutoInitEnabled": { _ in
                let enabled = try await provider.getAutoInitEnabled()
                return FCMAutoInit(enabled: enabled).toPayload()
            },
            "setAutoInitEnabled": { payload in
                guard case let .bool(enabled) = payload["enabled"] else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: enabled")
                }
                try await provider.setAutoInitEnabled(enabled)
                return [:]
            },
        ]
    }
}

File: android/push-fcm/src/main/kotlin/com/devryner/rynbridge/push/fcm/PushFCMModule.kt
package com.devryner.rynbridge.push.fcm

import com.devryner.rynbridge.core.*

class PushFCMModule(provider: PushFCMProvider) : BridgeModule {

    override val name = "push-fcm"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "getToken" to { _ ->
            val token = provider.getToken()
            FCMToken(token = token).toPayload()
        },
        "deleteToken" to { _ ->
            provider.deleteToken()
            emptyMap()
        },
        "subscribeToTopic" to { payload ->
            val topic = payload["topic"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: topic")
            provider.subscribeToTopic(topic)
            emptyMap()
        },
        "unsubscribeFromTopic" to { payload ->
            val topic = payload["topic"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: topic")
            provider.unsubscribeFromTopic(topic)
            emptyMap()
        },
        "getAutoInitEnabled" to { _ ->
            val enabled = provider.getAutoInitEnabled()
            FCMAutoInit(enabled = enabled).toPayload()
        },
        "setAutoInitEnabled" to { payload ->
            val enabled = payload["enabled"]?.boolValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: enabled")
            provider.setAutoInitEnabled(enabled)
            emptyMap()
        }
    )
}
